modules/apps/layout/layout-seo-web/src/main/java/com/liferay/layout/seo/web/internal/util/AssetDisplayPageFriendlyURLMapper.java

```diff
@@ -0,0 +1,137 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.layout.seo.web.internal.util;
+
+import com.liferay.asset.kernel.AssetRendererFactoryRegistryUtil;
+import com.liferay.asset.kernel.model.AssetEntry;
+import com.liferay.asset.kernel.model.AssetRenderer;
+import com.liferay.asset.kernel.model.AssetRendererFactory;
+import com.liferay.journal.constants.JournalArticleConstants;
+import com.liferay.journal.model.JournalArticle;
+import com.liferay.petra.string.CharPool;
+import com.liferay.portal.kernel.exception.PortalException;
+import com.liferay.portal.kernel.model.Layout;
+import com.liferay.portal.kernel.util.StringUtil;
+import com.liferay.portal.kernel.util.Validator;
+import com.liferay.portal.kernel.util.WebKeys;
+
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+
+import javax.servlet.http.HttpServletRequest;
+
+/**
+ * @author Adolfo Pérez
+ */
+public class AssetDisplayPageFriendlyURLMapper {
+
+	public static AssetDisplayPageFriendlyURLMapper create(
+			HttpServletRequest httpServletRequest)
+		throws PortalException {
+
+		Layout layout = (Layout)httpServletRequest.getAttribute(WebKeys.LAYOUT);
+
+		if ((layout != null) && !layout.isContentDisplayPage()) {
+			return new AssetDisplayPageFriendlyURLMapper(null);
+		}
+
+		AssetEntry assetEntry = (AssetEntry)httpServletRequest.getAttribute(
+			WebKeys.LAYOUT_ASSET_ENTRY);
+
+		if (Objects.equals(
+				assetEntry.getClassName(), JournalArticle.class.getName())) {
+
+			AssetRendererFactory<?> assetRendererFactory =
+				AssetRendererFactoryRegistryUtil.
+					getAssetRendererFactoryByClassNameId(
+						assetEntry.getClassNameId());
+
+			AssetRenderer<?> assetRenderer =
+				assetRendererFactory.getAssetRenderer(assetEntry.getClassPK());
+
+			return new AssetDisplayPageFriendlyURLMapper(
+				(JournalArticle)assetRenderer.getAssetObject());
+		}
+
+		return new AssetDisplayPageFriendlyURLMapper(null);
+	}
+
+	public String getMappedAssetDisplayPageFriendlyURL(
```

Comment:
Maybe the AssetDisplayPageFriendlyURLProvider could be helpful with this?

Reply:
Mmm, but that does only work for the current locale, right? We need to process the URL for all available locales.